File: kaji/packages/py/src/backends/postgres/store.py
"""Postgres implementation of the durable event-store protocol."""

from __future__ import annotations

from collections.abc import AsyncIterator
from contextlib import asynccontextmanager
import json
from typing import Any

from kaji.events.errors import EventIdConflictError
from kaji.events.schemas import (
    NewKajiEvent,
    StoredKajiEvent,
    revalidate_new_event,
    revalidate_stored_event,
)
from kaji.events.store.base import AppendResult, prepare_stored_event


def _get_psycopg() -> Any:
    """Import the optional Postgres driver only when this adapter is used."""

    try:
        import psycopg  # noqa: PLC0415

        return psycopg
    except ImportError as exc:
        raise ImportError(
            "Postgres support requires `pip install 'kaji[postgres]'`."
        ) from exc


def _draft_payload(event: NewKajiEvent | StoredKajiEvent) -> dict[str, Any]:
    return event.model_dump(mode="json", exclude={"sequence"})


def _decode_event(value: Any) -> StoredKajiEvent:
    if isinstance(value, str):
        value = json.loads(value)
    return revalidate_stored_event(value)


class PostgresEventStore:
    """Durable append-only events using a transactional per-session counter."""

    def __init__(self, dsn: str) -> None:
        if not isinstance(dsn, str) or not dsn.strip():
            raise TypeError("dsn must be a non-empty string")
        self._dsn = dsn

    @asynccontextmanager
    async def _connection(self) -> AsyncIterator[Any]:
        psycopg = _get_psycopg()
        connection = await psycopg.AsyncConnection.connect(self._dsn)
        try:
            yield connection
        finally:
            await connection.close()

    async def _existing(self, event_id: str) -> StoredKajiEvent | None:
        async with self._connection() as connection:
            row = await (
                await connection.execute(
                    "SELECT event_json FROM kaji_events WHERE event_id = %s",
                    (event_id,),
                )
            ).fetchone()
            return None if row is None else _decode_event(row[0])

    async def append(self, event: NewKajiEvent) -> AppendResult:
        draft = revalidate_new_event(event)
        try:
            async with self._connection() as connection:
                async with connection.transaction():
                    row = await (
                        await connection.execute(
                            """
                            INSERT INTO kaji_event_sequences (session_id, next_sequence)
                            VALUES (%s, 2)
                            ON CONFLICT (session_id)
                            DO UPDATE SET next_sequence = kaji_event_sequences.next_sequence + 1
                            RETURNING next_sequence - 1
                            """,
                            (draft.session_id,),
                        )
                    ).fetchone()
                    assert row is not None
                    stored = prepare_stored_event(draft, int(row[0]))
                    await connection.execute(
                        """
                        INSERT INTO kaji_events (session_id, sequence, event_id, event_json)
                        VALUES (%s, %s, %s, %s::jsonb)
                        """,
                        (
                            stored.session_id,
                            stored.sequence,
                            stored.id,
                            json.dumps(
                                stored.model_dump(mode="json"), separators=(",", ":")
                            ),
                        ),
                    )
                    return AppendResult(event=stored, inserted=True)
        except Exception as exc:
            if getattr(exc, "sqlstate", None) != "23505":
                raise
            existing = await self._existing(draft.id)
            if existing is not None and _draft_payload(existing) == _draft_payload(
                draft
            ):
                return AppendResult(event=existing, inserted=False)
            raise EventIdConflictError(draft.id) from exc

    async def get_events(
        self,
        session_id: str,
        *,
        after_sequence: int = 0,
        limit: int | None = None,
    ) -> list[StoredKajiEvent]:
        if after_sequence < 0:
            raise ValueError("after_sequence must be non-negative")
        if limit is not None and limit < 0:
            raise ValueError("limit must be non-negative")
        if limit == 0:
            return []
        query = (
            "SELECT event_json FROM kaji_events "
            "WHERE session_id = %s AND sequence > %s ORDER BY sequence"
        )
        parameters: tuple[Any, ...] = (session_id, after_sequence)
        if limit is not None:
            query += " LIMIT %s"
            parameters += (limit,)
        async with self._connection() as connection:
            rows = await (await connection.execute(query, parameters)).fetchall()
        return [_decode_event(row[0]) for row in rows]

    async def last_sequence(self, session_id: str) -> int:
        async with self._connection() as connection:
            row = await (
                await connection.execute(
                    """
                    SELECT sequence FROM kaji_events
                    WHERE session_id = %s ORDER BY sequence DESC LIMIT 1
                    """,
                    (session_id,),
                )
            ).fetchone()
        return 0 if row is None else int(row[0])

    async def purge_session(self, session_id: str) -> bool:
        if not isinstance(session_id, str) or not session_id.strip():
            raise TypeError("session_id must be a non-empty string")
        async with self._connection() as connection:
            async with connection.transaction():
                deleted = await connection.execute(
                    "DELETE FROM kaji_events WHERE session_id = %s",
                    (session_id,),
                )
                await connection.execute(
                    "DELETE FROM kaji_event_sequences WHERE session_id = %s",
                    (session_id,),
                )
        return deleted.rowcount > 0

File: kaji/packages/ts/src/backends/postgres/store.ts
import { EventIdConflictError } from "@/events/errors";
import { structurallyEqualJson } from "@/events/json";
import {
  type NewKajiEvent,
  type StoredKajiEvent,
  snapshotNewEvent,
  snapshotStoredEventForAppend,
  validateStoredEvent,
} from "@/events/schemas";
import type { AppendResult, EventStore } from "@/events/store";

type Sql = any;

function draftOf(event: NewKajiEvent | StoredKajiEvent): unknown {
  if (!("sequence" in event)) return event;
  const { sequence: _, ...draft } = event;
  return draft;
}

function decodeEvent(value: unknown): StoredKajiEvent {
  return validateStoredEvent(typeof value === "string" ? JSON.parse(value) : value);
}

export class PostgresEventStore implements EventStore {
  private sql: Sql | undefined;

  constructor(private readonly url: string) {
    if (url.trim().length === 0) throw new TypeError("url must be a non-empty string");
  }

  private async client(): Promise<Sql> {
    if (this.sql === undefined) {
      const { default: postgres } = await import("postgres");
      this.sql = postgres(this.url);
    }
    return this.sql;
  }

  private async existing(eventId: string): Promise<StoredKajiEvent | undefined> {
    const sql = await this.client();
    const rows = (await sql`
      SELECT event_json FROM kaji_events WHERE event_id = ${eventId}
    `) as Array<{ event_json: unknown }>;
    return rows[0] === undefined ? undefined : decodeEvent(rows[0].event_json);
  }

  async append(input: NewKajiEvent): Promise<AppendResult> {
    const event = snapshotNewEvent(input);
    const sql = await this.client();
    try {
      return await sql.begin(async (transaction: Sql) => {
        const [sequenceRow] = (await transaction`
          INSERT INTO kaji_event_sequences (session_id, next_sequence)
          VALUES (${event.session_id}, 2)
          ON CONFLICT (session_id)
          DO UPDATE SET next_sequence = kaji_event_sequences.next_sequence + 1
          RETURNING next_sequence - 1 AS next_sequence
        `) as Array<{ next_sequence: number }>;
        if (sequenceRow === undefined) throw new Error("sequence allocation returned no row");
        const stored = snapshotStoredEventForAppend({
          ...event,
          sequence: Number(sequenceRow.next_sequence),
        });
        await transaction`
          INSERT INTO kaji_events (session_id, sequence, event_id, event_json)
          VALUES (
            ${stored.session_id},
            ${stored.sequence},
            ${stored.id},
            ${JSON.stringify(stored)}::jsonb
          )
        `;
        return { event: stored, inserted: true };
      });
    } catch (error) {
      if ((error as { code?: string }).code !== "23505") throw error;
      const existing = await this.existing(event.id);
      if (existing !== undefined && structurallyEqualJson(draftOf(existing), event)) {
        return { event: existing, inserted: false };
      }
      throw new EventIdConflictError(event.id);
    }
  }

  async getEvents(
    sessionId: string,
    options: { afterSequence?: number; limit?: number } = {},
  ): Promise<StoredKajiEvent[]> {
    const afterSequence = options.afterSequence ?? 0;
    const limit = options.limit;
    if (!Number.isInteger(afterSequence) || afterSequence < 0) {
      throw new RangeError("afterSequence must be a non-negative integer");
    }
    if (limit !== undefined && (!Number.isInteger(limit) || limit < 0)) {
      throw new RangeError("limit must be a non-negative integer");
    }
    if (limit === 0) return [];
    const sql = await this.client();
    const rows =
      limit === undefined
        ? ((await sql`
            SELECT event_json FROM kaji_events
            WHERE session_id = ${sessionId} AND sequence > ${afterSequence}
            ORDER BY sequence
          `) as Array<{ event_json: unknown }>)
        : ((await sql`
            SELECT event_json FROM kaji_events
            WHERE session_id = ${sessionId} AND sequence > ${afterSequence}
            ORDER BY sequence LIMIT ${limit}
          `) as Array<{ event_json: unknown }>);
    return rows.map((row) => decodeEvent(row.event_json));
  }

  async lastSequence(sessionId: string): Promise<number> {
    const sql = await this.client();
    const rows = (await sql`
      SELECT sequence FROM kaji_events
      WHERE session_id = ${sessionId} ORDER BY sequence DESC LIMIT 1
    `) as Array<{ sequence: number }>;
    return rows[0] === undefined ? 0 : Number(rows[0].sequence);
  }

  async purgeSession(sessionId: string): Promise<boolean> {
    if (sessionId.trim().length === 0) throw new TypeError("sessionId must be a non-empty string");
    const sql = await this.client();
    return sql.begin(async (transaction: Sql) => {
      const deleted = await transaction`
        DELETE FROM kaji_events WHERE session_id = ${sessionId}
      `;
      await transaction`DELETE FROM kaji_event_sequences WHERE session_id = ${sessionId}`;
      return deleted.count > 0;
    });
  }

  async close(): Promise<void> {
    await this.sql?.end({ timeout: 5 });
    this.sql = undefined;
  }
}
